### PR TITLE
EE 9765 Alias Surname Combinations Generator

### PR DESCRIPTION
**Pull request by owainrobertsbjss**
Added a new NameMatchingCandidatesGenerator called AliasSurnameCombinations which produces the combinations of names shown in EE-9244.

I have intentionally not wired it into the rest of the code as the BDD is not ready and there may be changes necessary for this that would confuse this PR.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinations.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public List<CandidateName> generateCandidates(InputNames inputNames) {
+        List<CandidateName> candidateNames = new ArrayList<>();
+
+        if (inputNames.aliasSurnames().isEmpty()) {
+            return candidateNames;
+        }
+
+        for (String firstName : inputNames.firstNames()) {
+            candidateNames.add(new CandidateName(firstName, inputNames.fullLastName()));
+        }
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinations.java]: I thought this was done by a different candidate generator?

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    private AliasSurnameCombinationsFunctions() {
+        throw new UnsupportedOperationException("Helper class for AliasSurnameCombinationsFunctions containing only static methods - no need to instantiate.");
+    }
+
+    static List<String> removeName(String nameToRemove, List<String> names) {
+        return names.stream()
+                .filter(name -> !name.equals(nameToRemove))
+                .collect(Collectors.toList());
+    }
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java]: Is this a good candidate for List.remove?

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    static List<CandidateName> nonAliasFirstAliasLastCombinations(List<String> nonAliasNames, List<String> aliasSurnames) {
+        List<CandidateName> candidateNames = new ArrayList<>();
+
+        List<String> reversedAliasSurnames = new ArrayList<>(aliasSurnames);
+        reverse(reversedAliasSurnames);
```
Comment (mikehalmaho) [inline src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java]: Is this method working from a side effect?  Feels like it should return a new list.

File: src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctionsTest.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                new CandidateName("Evans", "Smith")
+        );
+
+        assertThat(nonAliasSurnameAsFirstNameCombinations(inputNames)).isEqualTo(expected);
+    }
+
+    @Test
+    public void nonAliasSurnameAsFirstNameCombinationsForTwoFirstNamesTwoLastNames() {
+        InputNames inputNames = new InputNames("John David", "Smith Evans");
+        List<CandidateName> expected = asList(
```
Comment (mikehalmaho) [inline src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctionsTest.java]: Though I don't understand the alias rules, it feels like this class should do nothing if there are no alias names?

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinations.java (hunk 1 above)
Comment (owainrobertsbjss) [reply src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinations.java]: It probably is but the ordering for alias matching is very specific and precedes all the multi-name, combinations and special character stuff.  This keeps the alias rules self-contained and deduplication will make sure we don't repeat.

Comment (owainrobertsbjss) [reply src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinations.java]: However it shouldn't be doing the full last name but rather each individual word in turn - an oversight that I'll correct.

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java (hunk 3 above)
Comment (owainrobertsbjss) [reply src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java]: Yes it's Collections.reverse. I probably should put the qualifier in to make this clear. I don't like it but I have yet to discover a more fluent way to reverse a list. Definitely open to suggestions!

File: src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctionsTest.java (hunk 4 above)
Comment (owainrobertsbjss) [reply src/test/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctionsTest.java]: The AliasSurnameCombinations does do nothing when there's no alias names. But it creates some combinations that don't involve alias names using some of the methods in its functions class.

File: src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java (hunk 2 above)
Comment (owainrobertsbjss) [reply src/main/java/uk/gov/digital/ho/pttg/application/namematching/candidates/AliasSurnameCombinationsFunctions.java]: Yes - I've made the change

**Pull request merged**